framework/test_app/runners/cloud_run/cloud_run_base_runner.py

```diff
@@ -0,0 +1,200 @@
+from framework.test_app.runners import base_runner
+
+logger = logging.getLogger(__name__)
+
+_HighlighterYaml = framework.helpers.highlighter.HighlighterYaml
+_helper_datetime = framework.helpers.datetime
+_datetime = dt.datetime
+_timedelta = dt.timedelta
+
+
+@dataclasses.dataclass(frozen=True)
+class RunHistory:
+    revision_id: str
+    time_start_requested: _datetime
+    time_start_completed: Optional[_datetime]
+    time_stopped: _datetime
+
+
+class CloudRunBaseRunner(base_runner.BaseRunner, metaclass=ABCMeta):
+    """Runs xDS Test Client/Server on Cloud Run."""
+
+    project: str
+    service_name: str
+    image_name: str
+    network: Optional[str] = None
+    tag: str = "latest"
+    region: str = "us-central1"
+    gcp_api_manager: gcp.api.GcpApiManager
+    current_revision: Optional[str] = None
+
+    run_history: collections.deque[RunHistory]
+
+    time_start_requested: Optional[dt.datetime] = None
+    time_start_completed: Optional[dt.datetime] = None
+    time_stopped: Optional[dt.datetime] = None
+
+    def __init__(
+        self,
+        project: str,
+        service_name: str,
+        image_name: str,
+        region: str,
+        gcp_api_manager: gcp.api.GcpApiManager,
+        network: Optional[str] = None,
+    ) -> None:
+        super().__init__()
+
+        self.project = project
+        self.service_name = service_name
+        self.image_name = image_name
+        self.network = network
+        self.region = region
+        self.current_revision = None
+        self.gcp_ui_url = gcp_api_manager.gcp_ui_url
+        self.gcp_api_manager = gcp_api_manager
+
+        # Persistent across many runs.
+        self.run_history = collections.deque()
+
+        # Mutable state associated with each run.
+        self._reset_state()
+
+        # Highlighter.
+        self._highlighter = _HighlighterYaml()
+
+        self._initalize_cloud_run_api_manager()
+
+    def _initalize_cloud_run_api_manager(self):
+        """Initializes the CloudRunV2."""
+        self.cloud_run_api_manager = cloud_run.CloudRunV2(
+            project=self.project,
+            region=self.region,
+            api_manager=self.gcp_api_manager,
+        )
+
+    def run(self, **kwargs):
+        if self.time_start_requested and not self.time_stopped:
+            if self.time_start_completed:
+                raise RuntimeError(
+                    f"Service {self.service_name}: has already been started "
+                    f"at {self.time_start_completed.isoformat()}"
+                )
+            raise RuntimeError(
+                f"Service {self.service_name}: start has already been "
+                f"requested at {self.time_start_requested.isoformat()}"
+            )
+
+        self._reset_state()
+        self.time_start_requested = _datetime.now()
+        self.current_revision = self.cloud_run_api_manager.deploy_service(
+            self.service_name,
+            self.image_name,
+        ).url
```

Comment:
let's move this to the cloud run server runner, and then

1. store the full service object, not just url - as `CloudRunServerRunner.service`
2. clean it up as a part of `CloudRunServerRunner.cleanup`

Reply:
note that `KubernetesBaseRunner.run` pretty much just deals with history tracking